### DISCUSSION TITLE
Package version questions

### DISCUSSION THREAD

**Discussion by wdfinch** [module-federation/module-federation-examples#769] [Q&A] [answered]
Hi all,

Thank you so much for all the work put into module federation! I migrated my company's entire frontend application to use it and it's been great. After setting it up and reading the book linked in this repo, I had a few questions that I can't seem to find an answer to. I would really appreciate if someone could point me in the right direction.

1. If we use the `singleton` option on a shared dependency and we have two different versions of a package present (ex. React 17.0.0 and React 17.0.01), then what would happen when loading the second version? Would we use the first library loaded be used, the higher version, etc.?

2. If we are using a workspace system and we have a private package in our workspace that is shared between multiple apps, then what would happen if multiple apps were using a different version of the library but the package.json version number was the same? So for example `myPackage` version 1.0.0 was used in app A and then I deploy app B but make a few changes to `myPackage`  without changing the version (still 1.0.0 ). Would webpack use a hash for the name of the package based on the contents to specify that they are different packages or just go off the version number?

3. What is the naming convention on build shared packages? I noticed that after a production build, packages have names like `4867.js`. Is this a hash on the contents of the package, version number, etc.? 

Thanks so much and really appreciate any help.

**Comment by wdfinch** [comment]
@ScriptedAlchemy do you have any insights here? I couldn't find this in the book. Would really appreciate it if you could point me in the right direction.

**Comment by ScriptedAlchemy** [comment, marked as answer]
If you're using ^ in package json. Webpack will upgrade to the latest. 

When / if you're using singletons with various versions hard locked. Webpack will likely go with the best (latest patch) available within one share scope

**Reply by wdfinch** [reply]
Thanks for the reply!
